mount/mount.go

```diff
@@ -189,6 +189,24 @@ func (d *Datastore) Put(key ds.Key, value []byte) error {
 	return cds.Put(k, value)
 }
 
+// Sync implements Datastore.Sync
+func (d *Datastore) Sync(prefix ds.Key) error {
+	// Sync all mount points below the prefix
+	// Sync the mount point right at (or above) the prefix
+	dstores, mountPts, rest := d.lookupAll(prefix)
+	for i, suffix := range rest {
+		if err := dstores[i].Sync(suffix); err != nil {
+			return err
+		}
+
+		if mountPts[i].Equal(prefix) || suffix.String() != "/" {
+			return nil
+		}
```

Comment:
Hm. This looks like a bug in `lookupAll`. `lookupAll` should only return mounts that can contain descendents of `key`.

Reply:
We can merge this as-is and fix it later or move this logic into `lookupAll` now.

Reply:
Could you clarify what you think the bug is here? It looks like the documented behavior for `lookupAll`.

Reply:
`lookupAll` is supposed to return all datastores that _might_ contain keys that are descendants of the given key. That means we should be calling `Sync` on all datastores returned by `lookupAll` datastores and we shouldn't have this extra check.

However, if we have `["/foo", "/foo/bar"]`, `"/foo"` should "cover" `"/foo/bar"` given that `lookup` will always return `"/foo"`. However, when given the prefix `"/foo"`, `lookupAll` will return _both_.

Note: this is only _technically_ a bug. Nobody would ever mount `"/foo"` over `"/foo/bar"`.

Reply:
> Nobody would ever mount "/foo" over "/foo/bar"

I thought it was expected behavior that we might mount "/" and also "/foo", or is the root a special case?

Reply:
My first example was bad: order matters. If we mount "/" over "/foo", "/" will win (I think? check me here). In go-ipfs, we mount "/blocks" over "/".

Reply:
I say "nobody would ever mount "/foo" over "/foo/bar" because "/foo" is a prefix of "/foo/bar" (and we check mounts in-order".

Reply:
I was wrong. We reverse-sort datastores so "/foo/bar" always comes before "/foo".

However, the bug was valid. Fix in #141.